airflow/models/taskinstance.py

```diff
@@ -2744,104 +2733,126 @@ def _run_raw_task(
-            ti.log.info("Creating event through asset name reference %r", asset_model.name)
+        asset_event_extras: dict[AssetUniqueKey, dict] = {
+            AssetUniqueKey(**event["dest_asset_key"]): event["extra"]
+            for event in outlet_events
+            if "source_alias_name" not in event
+        }
+
+        bad_asset_keys: set[AssetUniqueKey | AssetNameRef | AssetUriRef] = set()
+
+        for key in asset_keys:
+            try:
+                am = asset_models[key]
+            except KeyError:
+                bad_asset_keys.add(key)
+                continue
+            ti.log.debug("register event for asset %s", am)
             asset_manager.register_asset_change(
                 task_instance=ti,
-                asset=asset_model,
-                extra=outlet_events[asset_model].extra,
+                asset=am,
+                extra=asset_event_extras.get(key),
                 session=session,
             )
-        asset_stmt = select(AssetModel).where(AssetModel.uri.in_(asset_uri_refs), AssetModel.active.has())
-        for asset_model in session.scalars(asset_stmt):
-            ti.log.info("Creating event for through asset URI reference %r", asset_model.uri)
-            asset_manager.register_asset_change(
-                task_instance=ti,
-                asset=asset_model,
-                extra=outlet_events[asset_model].extra,
+
+        if asset_name_refs:
+            asset_models_by_name = {key.name: am for key, am in asset_models.items()}
+            asset_event_extras_by_name = {key.name: extra for key, extra in asset_event_extras.items()}
+            for nref in asset_name_refs:
+                try:
+                    am = asset_models_by_name[nref.name]
+                except KeyError:
+                    bad_asset_keys.add(nref)
+                    continue
+                ti.log.debug("register event for asset name ref %s", am)
+                asset_manager.register_asset_change(
+                    task_instance=ti,
+                    asset=am,
+                    extra=asset_event_extras_by_name.get(nref.name),
+                    session=session,
+                )
+        if asset_uri_refs:
+            asset_models_by_uri = {key.uri: am for key, am in asset_models.items()}
+            asset_event_extras_by_uri = {key.uri: extra for key, extra in asset_event_extras.items()}
+            for uref in asset_uri_refs:
+                try:
+                    am = asset_models_by_uri[uref.uri]
+                except KeyError:
+                    bad_asset_keys.add(uref)
+                    continue
+                ti.log.debug("register event for asset uri ref %s", am)
+                asset_manager.register_asset_change(
+                    task_instance=ti,
+                    asset=am,
+                    extra=asset_event_extras_by_uri.get(uref.uri),
+                    session=session,
+                )
+
+        if bad_asset_keys:
```

Comment:
if we check the condition here, does that mean we want handle good alias when bad asset key exists

Reply:
Not following

Reply:
If my understanding is correct, here is what we trying to do

1. collect assets and register the good ones
2. see if there are bad assets, if so, raise an exception
3. collect asset aliases and register the good ones
4. see if there are bad asset aliases; if so, raise an exception

so my question is "will it stop at step 2 while we probably should also run step 3"

Reply:
In practice 2 can’t fire unless you send a faulty request. I guess I can move it to the last instead? (Or before 4, but it’s easier to move it to the last since 3 and 4 are in the same if block.)

Reply:
yep, maybe let's move it to the last. 👍 thanks!